Tools/BlockTypePaletteGenerator/UpgradeGenerator.lua

```diff
@@ -0,0 +1,223 @@
+	local flatName = ""
+	local propKey = ""
+	aFullFlatName:gsub("([^:]+)",
+		function (aPart)
+			if (numParts == 0) then
+				flatName = aPart
+			elseif (numParts == 1) then
+				flatName = flatName .. ":" .. aPart
+			elseif (numParts % 2 == 0) then
+				propKey = aPart
+			else
+				props[propKey] = aPart
+			end
+			numParts = numParts + 1
+		end
+	)
+	return flatName, props
+end
+
+
+
+
+
+--- Returns the minecraft block name, created from the flat name if present, or synthesized
+-- from the Minutor name
+-- If the flat name contains encoded block properties, it returns those properties as a dict-table
+-- in the second return value
+local function processBlockName(aFlatName, aMinutorName)
+	if (aFlatName) then
+		assert(type(aFlatName) == "string")
+		return splitFlatName(aFlatName)
+	end
+	if not(type(aMinutorName) == "string") then
+		return nil
+	end
+	return "minecraft:" .. (aMinutorName:lower():gsub(" ", "_")), {}
+end
+
+
+
+
+
+
+--- Serializes the properties from the JSON / array table format into a single output string
+-- Concatenates all properties with \t as the delimiting character
+local function serializeProperties(aProperties)
+	local res = {}
+	local idx = 1
+	for k, v in pairs(aProperties or {}) do
+		res[idx] = k
+		res[idx + 1] = v
+		idx = idx + 2
+	end
+	return table.concat(res, "\t")
+end
+
+
+
+
+
+--- Parses the vanilla_ids.json into a common registry format
+-- The returned registry is an array-table of
+-- {blockType = 1, blockMeta = 2, blockTypeName = "name", properties = {key = value, ...}}
+local function parseRegistry(aJsonString)
+	assert(type(aJsonString) == "string")
+
+	-- Parse the JSON:
+	local lj = require("lunajson")
+	local input = lj.decode(aJsonString)
+	if (not(input) or (input["type"] ~= "block") or not(input["data"])) then
+		error("The input file doesn't contain vanilla IDs.")
+	end
+
+	-- Create the registry:
+	local registry = {}
+	local idx = 1
+	for _, entry in pairs(input["data"]) do
+		local id = entry["id"]
+		local parentBlockTypeName, props = processBlockName(entry["flatname"], entry["name"])
+		registry[idx] =
+		{
+			blockType = id,
+			blockMeta = 0,
+			blockTypeName = parentBlockTypeName,
+			properties = props,
+		}
+		idx = idx + 1
+		for _, variant in pairs(entry["variants"] or {}) do
+			local blockTypeName, props = processBlockName(variant["flatname"], variant["name"])
+			if not(blockTypeName) then
+				-- Some blocks don't have all their variants named ("brown mushroom block"), use the parent name in such a case
+				blockTypeName = parentBlockTypeName
+			end
+			registry[idx] =
+			{
+				blockType = id,
+				blockMeta = variant["data"],
+				blockTypeName = blockTypeName,
+				properties = props,
+			}
+			idx = idx + 1
+		end
+	end
+	return registry
+end
+
+
+
+
+
+--- Returns the prefix that is common for all block type names in the registry
+-- aRegistry is the parsed registry, as returned from parseRegistry()
+local function findCommonPrefix(aRegistryTable)
```

Comment:
I see some of these functions are in `UpgradeGenerator.lua` and `Generator.lua`. Is it an idea to put those in a different file so they can be shared?

Reply:
Not worth the trouble for just a handful of functions. The scripts will most likely never be touched again, only run against newer data files from newer minecraft versions.